File: DatabaseTypes/Permissions.fs
namespace DatabaseTypes

open System.Text.Json.Serialization
open DatabaseTypes.Identificators

type Permissions = {
    [<JsonPropertyName("owner_id")>]
    OwnerId : UserId
    [<JsonPropertyName("groups")>]
    Groups: List<GroupAccess>
    [<JsonPropertyName("members")>]
    Members: List<Member>
    [<JsonPropertyName("is_public")>]
    IsPublic: bool
}

File: react-ui/src/models/Permissions.ts
import { UserId, GroupId } from "./Identificators";

export interface Access {
    generate: boolean;
    read: boolean;
    write: boolean;
    admin: boolean;
}

export interface Member {
    user_id: UserId;
    name: string;
    access: Access;
}

export interface Group {
    id: GroupId;
    owner_id: UserId;
    name: string;
    description: string;
    members: Array<Member>;
}

export interface GroupAccess {
    group_id: GroupId;
    name: string;
    access: Access;
}

export interface Permissions {
    owner_id: UserId;
    groups: Array<GroupAccess>;
    members: Array<Member>;
}